File: fixtures/scale/julia/medium/src/Tokens.jl
module Tokens

export Token, Cursor, peek_token, next_token

struct Token
    type::Symbol
    value::String
end

mutable struct Cursor
    tokens::Vector{Token}
    pos::Int
end

Cursor(tokens::Vector{Token}) = Cursor(tokens, 1)

function peek_token(cursor::Cursor)
    return cursor.pos > length(cursor.tokens) ? nothing : cursor.tokens[cursor.pos]
end

function next_token(cursor::Cursor)
    token = peek_token(cursor)
    cursor.pos += 1
    return token
end

end

File: fixtures/scale/r/medium/lexer/tokens.r
# Token model: a mutable cursor over scanned tokens.

token_new <- function(type, value) {
  list(type = type, value = value)
}

token_stream <- function(tokens) {
  stream <- new.env()
  stream$tokens <- tokens
  stream$pos <- 1L
  stream
}

stream_peek <- function(stream) {
  if (stream$pos > length(stream$tokens)) NULL else stream$tokens[[stream$pos]]
}

stream_next <- function(stream) {
  token <- stream_peek(stream)
  stream$pos <- stream$pos + 1L
  token
}
